### ISSUE TITLE
Raw H264 stream: problem with framerate

### ISSUE THREAD

**Issue by raffaeler** [Microsoft/FFmpegInterop#165]
Hi, I can stream out of a RPi the raw data of an H264 video.
If I want to obtain a well-formed MP4 file, I would save the video as H264 and then wrap into an MP4 container:
- `ffmpeg -r 25 -i raw_video.h264 -vcodec copy wrapped_video.mp4`
or
- `MP4Box -add raw_video.h264 wrapped_video.mp4`
It works, but this is of course not possible when streaming.

If I save the raw H264 and play it with CreateFFmpegInteropMSSFromStream, the video is super-fast. I suppose the problem is due to the missing framerate.
From the ffmpeg rtsp help I don't see any useful options to make it work.

Suggestions?
TIA

**Comment by khouzam** [comment]
Can you try with the code from pull request #163?

We have an issue with small samples that might address the fast playback.

**Comment by raffaeler** [comment]
I cloned the repo, pulled the #163, recompiled the Win10 solution and run the C# example.
When I load the H264 raw video file, the result is:
`MF_MEDIA_ENGINE_ERR_DECODE : HRESULT - 0xC00D3E84`
I never had this error in my previous tests ... do you have any idea?

Using another H264 file I can see the first frame but the video doesn't play.
Is there some ffmpeg option to pass in the propertyset to support the H264 raw data?

**Comment by raffaeler** [comment]
Hi @khouzam thanks again for your support.
I did other tests with my own UWP app using the modified control but the problem is exactly the same.
Do you have any other suggestions?

**Comment by raffaeler** [comment]
Hi @khouzam sorry for pinging, but I was wondering if you have any solution for this problem.

Thanks

**Comment by khouzam** [comment]
Hi @raffaeler 

Is there a repro I can run on my side to see what's going on?

**Comment by raffaeler** [comment]
Would you like me to send a raw video captured with the default Raspberry PI utility?
It is not "processed" in any way, just raw H264 stream saved on disk.
I am trying to show it as a stream without encapsulate in the MP4 container.

**Comment by khouzam** [comment]
Yes, please, anything that can replicate the issues that you are seeing.

**Comment by raffaeler** [comment]
Hi @khouzam here is a video captured as a raw stream.
This specific video can be opened with VNC (this isn't always true). But if I try to watch it with FFmpegInterop, it plays in about 1 second instead its duration is about 12 seconds.
[raw-video.zip](https://github.com/Microsoft/FFmpegInterop/files/1271985/raw-video.zip)
Please let me know if you need another one (different length, etc.)
Thank you

**Comment by khouzam** [comment]
Ok, I've found the issue. Because this is a raw stream, the timestamps for the samples aren't and we're not handling this case properly. It should be a simple fix, because the samples do have duration, so we'd just have to keep increasing the timestamps for the current position based on the sample durations

**Comment by raffaeler** [comment]
Awesome. Do you believe to bea ble to make the fix before Wed? If I have time I would like to show the streaming in my demo @ Ignite next week.

**Comment by khouzam** [comment]
Here you go, please try the fix on https://github.com/Microsoft/FFmpegInterop/tree/gillesk/fixNoPts.

**Comment by raffaeler** [comment]
I just fetched the branch, recompiled everytyhing and it works!
I will make other tests but it looks amazing.
Thank you

**Comment by khouzam** [comment]
I've created Pull Request #171 to address the issue. As mentioned above, we'll keep track of the current presentation time based on the timestamp and the duration of samples as we're playing a stream and if a sample doesn't have a timestamp, it will infer the timestamp from the running timestamp.

**Comment by raffaeler** [comment]
@khouzam how many samples do you need to make the CreateFFmpegInteropMSSFromStream return a valid decoder?
I modified my sample in order to read the bytes from a websocket but CreateFFmpegInteropMSSFromStream returns null.

TIA

**Comment by khouzam** [comment]
I'm not sure about this. It will all depend on the how much data FFmpeg needs to identify the stream, this is not something that FFmpegInterop is doing. On the other hand, we're creating a wrapper on the stream, perhaps this is what is failing. Would it be possible to debug into this?

**Comment by raffaeler** [comment]
Yes I identified the problem two minutes ago.
The H264 stream begins with 00 00 00 01. If the stream already started, FFmpegInterop does not understand the 'random' sequence of bytes and failed.
I expected instead that it continue to read data (and discard) until finding the magic sequence.
If you want to repro, just remove the first 32 bytes from the files I sent you ... this way you simulate listening to a stream from the network.

BTW I am trying right now to identify the magic sequence and see if there is anything else to do.

**Comment by raffaeler** [comment]
@khouzam  Apparently this is not enough. Even after recognizing the magic bytes sequence and feeding the decode with that, the factory method returns null.
I am not an expert of H264 therefore I don't know if the beginning of the sequence (the beginning of the file in your case) is special.

**Comment by khouzam** [comment]
I tried removing random bytes from the raw file and it keeps working. I'm guessing that FFmpegInterop.cpp@205 is failing this call:
```
		// Convert asynchronous IRandomAccessStream to synchronous IStream. This API requires shcore.h and shcore.lib
		hr = CreateStreamOverRandomAccessStream(reinterpret_cast<IUnknown*>(stream), IID_PPV_ARGS(&fileStreamData));
```

The web socket is most likely not castable to a RandomAccessStream.

**Comment by raffaeler** [comment]
@khouzam
After recognizing the correct byte[] sequence, now I get a valid decoder,  but still doesn't work.
My steps are:
- I create an InMemoryRandomAccessStream as an intermediary buffer since the FFmpegInterpo need an IRandomAccessStream.
- I pull the data correctly from the websocket
- I write the data to the memory stream using the output stream obtained (before looping) with GetOutputStreamAt(0) 
But now the FFmpegInterop does not pull from my memory stream (is position is always 0).

**Comment by raffaeler** [comment]
@khouzam how many bytes do you need in the stream before rendering the image?
I see that if I buffer the data coming from the websocket, I can see an image (just the first frame) but then it fails with a weird c++ exception as if it is looking for "future" data that is not still available.
If the stream is a file stream, you have all the data at once, but processing from the network is of course very different...

**Comment by raffaeler** [comment]
@khouzam do you believe it is possible to have a short chat together to explain what is not working now.
The goal is very close but the live stream still doesn't work.
TIA

**Comment by khouzam** [comment]
Sure,

Can you send me an email and we can set something up.

**Comment by liujingsheng** [comment]
@khouzam 
I got a same problem as raffaeler, we have a media server wrap the raw h264 to flv and stream out through the websocket. 
no we want play it with uwp app, and use the ffmpeginterop with the same problem.
the mediaelement control only can play the stream for few seconds when i change the maxdelay to 1000.
then the mediaelement stoped play, but websocket is still working on receive data.

             _ws = new StreamWebSocket();
            await _ws.ConnectAsync(new Uri(wsuri.Text, UriKind.Absolute));     
            var stream = new Windows.Storage.Streams.InMemoryRandomAccessStream();
            var output = stream.GetOutputStreamAt(0);
            var writer = new DataWriter(output);
            int maxdelay = 1000;
            int delay = maxdelay;
            int index = 0;
            bool isplayed = false;
            DataReader reader = new DataReader(_ws.InputStream);
            do
            {
                var loaded = await reader.LoadAsync(10240);
                var buf = reader.ReadBuffer(loaded);
                writer.WriteBuffer(buf);          
                await writer.FlushAsync();
                await writer.StoreAsync();
 if (delay > 0)
                {
                    delay--;
                    Debug.Write("B");
              
                    continue;
                }
                // //delay = maxdelay;

                //Debug.Write(".");
                Debug.WriteLine(media.CurrentState);
                try
                {
                    //await output.WriteAsync(buf);
                   delay = maxdelay;
                    isCreated = true;
                        //stream.Seek((ulong)0);
                        var decoder = FFmpegInterop.FFmpegInteropMSS.CreateFFmpegInteropMSSFromStream(
                            stream, false, false, options);
                        if (decoder == null)
                        {
                            Debug.WriteLine("decoder was null, will retry");
                            continue;
                        }

                        var source = decoder.GetMediaStreamSource();
                        media.RealTimePlayback = true;            
                        media.SetMediaStreamSource(source);
                        media.Play();
                   
                }
   catch (Exception err)
                {
                    Debug.WriteLine(err.ToString());
                    stream.Dispose();
                    reader.Dispose();
                    return;
                }
             
            }
            while (true);

**Comment by khouzam** [comment]
Creating an InMemoryRandomAccessStream will not work as it will only copy the current portion of the data. There needs to be a wrapper around the WebSocket that provides the stream interface to the MediaStreamSource

**Comment by khouzam** [comment]
After thinking about it some more, the easiest thing that might be done would be to mimic the way we wrap the RandomAccessStream into an IStream interface and wrap the WebSocket inputstream. We could then have a CreateMediaStreamSource taking a WebSocket as an input parameter.

I haven't tried this but it should be the easiest step forward.

**Comment by liujingsheng** [comment]
I have try the way not just as FFmegainterop does, return a sample from websocket. and set a mediasource to mediaelement, it works.
but if i want use the FFmegainterop, is it possible.
I don't understand what your meaning about to mimic the way we wrap the RandomAccessStream into an IStream interface and wrap the WebSocket inputstream. 
I try to change the parameter to IIputstream of FFmpegInteropMSS::CreateMediaStreamSource()
but i see you cast the RandomAccessStream to Istream.
		hr = CreateStreamOverRandomAccessStream(reinterpret_cast<IUnknown*>(stream), IID_PPV_ARGS(&fileStreamData));

how can i Cast the IInputstream to a istream?

**Comment by raffaeler** [comment]
@liujingsheng I did not understand if @khouzam are asking us to make a test, or if instead he is anticipating a change for the ffmpeginterop codebase that he will do.

@khouzam could you please clarify?

TIA

**Comment by raffaeler** [comment]
@khouzam do you have any update? Thanks :)

**Comment by lukasf** [comment]
@raffaeler Can't you just use CreateFFmpegInteropMSSFromUri? This is how streaming scenarios are usually handled. If you can't use that, you need to add overloads to FFmpegInterop which take IInputStream. You must provide a custom implementation for FileStreamRead which directly works on IInputStream (it can't be wrapped to IStream). Pass NULL as FileStreamSeek in avio_alloc_context, since the stream cannot be seeked.

**Comment by raffaeler** [comment]
@lukasf I can't because I am streaming using websockets from the RPi.
But as you can read from the previous posts, I am already trying to read from the stream, but this is not enough:
1. the ffmpeginterop does not identify the h364 key frame. So I decoded manually the frames and start feeding the component with the h264 as I see the key frames.
2. even after that, the ffmpeginterop tries to read more than is currently available from the stream, and then hangs. I can therefore see few images and then stops.

I published the example and I am not alone with this problem, therefore I believe there is a problem in the ffmpeginterop component.

**Comment by lukasf** [comment]
You absolutely cannot use InMemoryRandomAccessStream for your scenario. As I wrote, you need to get the IInputStream from your WebSocket. Then have ffmpeg read directly from that IInputStream, by providing a custom FileStreamRead implementation. Only this will make sure that ffmpeg will not read beyond what is available.

**Comment by mcosmin222** [comment]
You have 2 choices, depending on how brave you are.

1) use the easy way lukasf describes.

2) create your own implementation of IRandomAccessStream wrapping an InputStream. This is much more complicated, but it guves you more flexibility in the long term.

Either way, when you run out of data you need to provide mock frames if the stream isnt over. I dont think ffmpeg interop supports this scebario.

**Comment by lukasf** [comment]
I do not see a need to provide mock frames. Every read call on a socket is delayed until at least one byte is available (or socket is closed). So if the socket does not provide data fast enough, playback will be delayed automatically. No need to patch mock frames into the stream (which could get horribly complicated).

Having FileStreamRead directly read the IInputStream is not only easier, but also more performant, compared to wrapping the IInputStream in a IRandomAccessStream, then wrapping that IRandomAccessStream in a COM IStream, then using a FileStreamRead to read form that IStream.

**Comment by mcosmin222** [comment]
Depending on the scenario and device, the system might enter a power saving mode or kill the app if frames are not provided in time. I had to deal with this scenario a few years back but there were tablets doing the playback on battery power.

You either need mock frames or call a method to notify progress on getting the next frame (i dont remember its name).

**Comment by lukasf** [comment]
You are talking about ReportSampleProgress.

https://docs.microsoft.com/en-us/uwp/api/windows.media.core.mediastreamsource#Windows_Media_Core_MediaStreamSource_SampleRequested

This could be needed if you expect longer interruptions (depends on the scenario).

If you are playing video (as seems to be the case here), I think you can also use DisplayRequest to keep display (and system) active as long as your are playing. The system will not enter standby as long as a DisplayRequest is active. This should be easier than the above method. But it does not work for pure audio playback (where you'd want display to go off but still keep the system on).

**Comment by raffaeler** [comment]
Streams on UWP are a kind of pain.
I will probably implement an InputStream, but it is embarassing seeing how far it would be easier with the classical streaming model in a desktop app.

**Comment by lukasf** [comment]
I generally like the async approach. But I agree that stream/data handling in UWP is overly complicated (and in some cases also pretty inefficient, forcing unneccessary copy operations). They could have made the API much easier and still async. 

Mixing async and non-async code is always a bit tricky. But then it's also not that hard. In c++/cx you can use create_task to get a task from the ReadAsync call result, then call task->get() to wait for and get the result. This will be blocking, but it's not a problem in our context.

**Comment by raffaeler** [comment]
I love the async approach too. But the API design is overcomplicated.
The shape on .net new networking libraries is far better and not too distant from the classical approach.
I studied, used and presented in conferences the async streaming and storage APIs in winrt, but they present multiple pitfalls, like the difficulties in implementing custom versions of stream and buffer objects.

**Comment by lukasf** [comment]
Yes, reading and writing data is very complicated. Especially the locked up IBuffer is hard to handle. It should be a one-liner to read and write binary data, from both managed and unmanaged code, as well as creating an IBuffer or InMemoryRandomAccessStream from existing data (without copy!).

**Comment by mcosmin222** [comment]
Buffer imutability helps prevent some typical buffer overflow attacks.

**Comment by lukasf** [comment]
Yeah, but it's not like they are really immutable. People who really need raw access will get it anyways. Hackers who want raw access will also get it. So I am not sure how much that really helps, given the cost of all that (in developer frustration, especially for beginners).

**Comment by raffaeler** [comment]
Buffers are never really immutable. The process memory is accessible by any code running in it or out of it, given the required privileges by the OS.
In one of my demos I show how the sandbox protects other processes from the sandboxed one. But also how easy is injecting code inside an UWP app and hijacking xaml or javascript UIs.
